src/safeposix/syscalls/net_calls.rs

```diff
@@ -1425,69 +1693,103 @@ impl Cage {
+                                                    "send",
+                                                    "The libc call to sendto failed!",
+                                                );
+                                            }
+                                            Err(()) => panic!(
+                                                "Unknown errno value from socket sendto returned!"
+                                            ),
+                                        };
+                                    } else {
+                                        return retval; //return the number of
+                                                       // bytes written to the
+                                                       // connected socket
+                                    }
                                 }
 
-                                //because socket must be connected it must have an inner socket
-                                let retval = sockhandle
-                                    .innersocket
-                                    .as_ref()
-                                    .unwrap()
-                                    .sendto(buf, buflen, None);
-                                if retval < 0 {
-                                    match Errno::from_discriminant(interface::get_errno()) {
-                                        Ok(i) => {
+                                //For INET sockets following the UDP protocol,
+                                //we don't need to check for connection status
+                                //as UDP is connection-less. This lets us grab
+                                //the remote address of the socket and send the
+                                //message in buf to it by calling sendto_syscall
+                                IPPROTO_UDP => {
+                                    let remoteaddr = match &sockhandle.remoteaddr {
+                                        Some(x) => x.clone(),
+                                        None => {
                                             return syscall_error(
-                                                i,
+                                                Errno::ENOTCONN,
                                                 "send",
-                                                "The libc call to sendto failed!",
+                                                "The descriptor is not connected",
                                             );
                                         }
-                                        Err(()) => panic!(
-                                            "Unknown errno value from socket sendto returned!"
-                                        ),
                                     };
-                                } else {
-                                    return retval;
+                                    //** Why is it necessary to drop here?? */
```

Comment:
in sendto_syscall, we need to acquire the fd/sockhandle with write/read lock again. If we do not release the lock here, deadlock would happen

Reply:
awesome! thanks